### PR TITLE
Properly generate 'nameof' expressions that bind correctly in the compiler

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
fixes https://github.com/dotnet/roslyn/issues/24212

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 1]
```diff
@@ -3201,9 +3201,11 @@ public override SyntaxNode AwaitExpression(SyntaxNode expression)
 
         public override SyntaxNode NameOfExpression(SyntaxNode expression)
         {
-            return this.InvocationExpression(
-                this.IdentifierName(CSharp.SyntaxFacts.GetText(SyntaxKind.NameOfKeyword)),
-                expression);
+            // A bit hacky, but we need to actually run ParseToken on the "nameof" text as there's
+            // no other way to get a token back that has the appropriate internal bit set that indicates
+            // this has the .ContextualKind of SyntaxKind.NameOfKeyword.
+            var nameofToken = SyntaxFactory.ParseToken("nameof");
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: ❔ Can we cache the result of this to a static field?

File: src/EditorFeatures/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeGeneration
+{
+    [UseExportProvider]
+    public class SyntaxGeneratorTests
+    {
+        [Fact]
+        public async Task TestNameOfBindsWithoutErrors()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs]: 💡 An additional test for a speculative semantic model would be good as well. This is a more likely scenario for analyzer authors to hit.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs]: will do.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: will do.

**Comment by CyrusNajmabadi** [comment]
@sharwell feedback accounted for.  Can you ptal?

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 3]
```diff
@@ -25,6 +25,13 @@ namespace Microsoft.CodeAnalysis.CSharp.CodeGeneration
     [ExportLanguageService(typeof(SyntaxGenerator), LanguageNames.CSharp), Shared]
     internal class CSharpSyntaxGenerator : SyntaxGenerator
     {
+        // A bit hacky, but we need to actually run ParseToken on the "nameof" text as there's no
+        // other way to get a token back that has the appropriate internal bit set that indicates
+        // this has the .ContextualKind of SyntaxKind.NameOfKeyword.
+        private static readonly IdentifierNameSyntax s_nameOfIdentifier =
+            SyntaxFactory.IdentifierName(SyntaxFactory.ParseToken("nameof"));
+
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]:
```suggestion
```

**Pull request merged**